Don't crash when UTF-8 conversion of filename in `fetchScriptFromLocalFileSystem` fails

danlliu opened the pull request:
#### a46b2777b0cb2e2c7076728045d6225eea1e483a
<pre>
Don&apos;t crash when UTF-8 conversion of filename in `fetchScriptFromLocalFileSystem` fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=281871">https://bugs.webkit.org/show_bug.cgi?id=281871</a>
<a href="https://rdar.apple.com/138178550">rdar://138178550</a>

Reviewed by Yijia Huang.

Previously, `fetchScriptFromLocalFileSystem` assumed that UTF-8 conversion of the filename
would always succeed, and had a `RELEASE_ASSERT` asserting such. However, if this conversion
fails for any reason, we&apos;ll trip the assertion. This patch cleans up this case by having the
program fail gracefully.

* Source/JavaScriptCore/jsc.cpp:
(fillBufferWithContentsOfFile):

Canonical link: <a href="https://commits.webkit.org/285721@main">https://commits.webkit.org/285721@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/786cc7e1e8222751f60564c9dd6884bf34fda4a2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73604 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53033 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26415 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77896 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24842 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62166 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/818 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57862 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16271 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48027 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63333 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/38268 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44615 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20818 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23175 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66741 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66354 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21167 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79492 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72861 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/921 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/394 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66233 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1063 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63342 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65514 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16192 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9371 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7554 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94643 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/885 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20801 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/914 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/901 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/920 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -1224,7 +1224,12 @@ static bool fillBufferWithContentsOfFile(FILE* file, Vector& buffer)
+        return false;
+    }
+    if (stat(fileNameUTF->data(), &statBuf) == -1) {
         fprintf(stderr, "Could not open file: %s\n", fileName.utf8().data());
```
Why not use fileNameUTF here as well? Same below.

kmiller68 replied:
Agreed.

hyjorc1 reviewed (approved):
r=me

annevk commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -1224,18 +1224,23 @@ static bool fillBufferWithContentsOfFile(FILE* file, Vector& buffer)
         return false;
     }
-    FILE* f = fopen(fileName.utf8().data(), "rb");
+    FILE* f = fopen(fileNameUTF->data(), "rb");
```
auto* maybe while here?

danlliu committed:
Don't crash when UTF-8 conversion of filename in `fetchScriptFromLocalFileSystem` fails
https://bugs.webkit.org/show_bug.cgi?id=281871
rdar://138178550

Reviewed by Yijia Huang.

Previously, `fetchScriptFromLocalFileSystem` assumed that UTF-8 conversion of the filename
would always succeed, and had a `RELEASE_ASSERT` asserting such. However, if this conversion
fails for any reason, we'll trip the assertion. This patch cleans up this case by having the
program fail gracefully.

* Source/JavaScriptCore/jsc.cpp:
(fillBufferWithContentsOfFile):

Canonical link: https://commits.webkit.org/285721@main
